patches/server/0936-Add-Entity-hidden-by-default-flag.patch

```diff
@@ -0,0 +1,207 @@
+@@ -2385,6 +2386,13 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
+                     flag = false;
+                 }
+                 // CraftBukkit end
++
++                // Paper start - Entity hiddenByDefault
++                if (this.entity.isHiddenByDefault() && !this.showToEvenWhenHiddenByDefault.contains(player.connection)) {
++                    flag = false;
++                }
++                // Paper end
++
+                 if (flag) {
+                     if (this.seenBy.add(player.connection)) {
+                         this.serverEntity.addPairing(player);
+diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
+index f925a8d550ecbf2044a37bfe58b30d6578c5f6af..dfda6e2f38be417f19dc9d7847f3cae80fb8fac2 100644
+--- a/src/main/java/net/minecraft/world/entity/Entity.java
++++ b/src/main/java/net/minecraft/world/entity/Entity.java
+@@ -398,6 +398,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
+     @javax.annotation.Nullable
+     private UUID originWorld;
+     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
++    public boolean hiddenByDefault; // Paper - Entity hiddenByDefault
+ 
+     public void setOrigin(@javax.annotation.Nonnull Location location) {
+         this.origin = location.toVector();
+@@ -4431,4 +4432,14 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
+         return ((ServerChunkCache) level.getChunkSource()).isPositionTicking(this);
+     }
+     // Paper end
++
++    // Paper start - Entity hiddenByDefault
++    public void setHiddenByDefault(boolean hiddenByDefault) {
++        this.hiddenByDefault = hiddenByDefault;
++    }
++
++    public boolean isHiddenByDefault() {
++        return this.hiddenByDefault;
++    }
++    // Paper end
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+index 6549ade8e19807c523e5a1dc68b66585aad438b1..0d3b808472a4ba25266aa3da1b27d91af775a267 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+@@ -1355,4 +1355,26 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+         return !this.getHandle().level.noCollision(this.getHandle(), aabb);
+     }
+     // Paper End - Collision API
++
++    // Paper start - Entity hiddenByDefault
++    @Override
++    public void setHiddenByDefault(boolean hiddenByDefault) {
++        this.getHandle().setHiddenByDefault(hiddenByDefault);
++
++        // We need to update the players again
++        ServerLevel world = ((CraftWorld) this.getWorld()).getHandle();
++        ChunkMap.TrackedEntity entityTracker = world.getChunkSource().chunkMap.entityMap.get(this.getEntityId());
```

Comment:
This can be fast-tracked using `getHandle().tracker`, saves you the world casting too.

Reply:
Thanks!